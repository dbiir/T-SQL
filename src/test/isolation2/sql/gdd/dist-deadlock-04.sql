DROP TABLE IF EXISTS t04a;
CREATE TABLE t04a (id int, val int);
INSERT INTO t04a (id, val) SELECT i, i FROM generate_series(1, 100) i;

DROP TABLE IF EXISTS t04b;

10: BEGIN;
20: BEGIN;

10: UPDATE t04a SET val=val WHERE id=segid(0,1);
10: CREATE TABLE t04b (id int);

20: UPDATE t04a SET val=val WHERE id=segid(1,1);

-- seg 1: con10 ==> con20, xid lock
10&: UPDATE t04a SET val=val WHERE id=segid(1,1);

-- seg -1: con20 ==> con10, xid lock
20>: CREATE TABLE t04b (id int);

-- con20 will be cancelled by gdd
20<:
20q:

-- no more deadlock
10<:
10q:
