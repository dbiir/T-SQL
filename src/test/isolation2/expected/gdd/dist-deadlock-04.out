DROP TABLE IF EXISTS t04a;
DROP
CREATE TABLE t04a (id int, val int);
CREATE
INSERT INTO t04a (id, val) SELECT i, i FROM generate_series(1, 100) i;
INSERT 100

DROP TABLE IF EXISTS t04b;
DROP

10: BEGIN;
BEGIN
20: BEGIN;
BEGIN

10: UPDATE t04a SET val=val WHERE id=segid(0,1);
UPDATE 1
10: CREATE TABLE t04b (id int);
CREATE

20: UPDATE t04a SET val=val WHERE id=segid(1,1);
UPDATE 1

-- seg 1: con10 ==> con20, xid lock
10&: UPDATE t04a SET val=val WHERE id=segid(1,1);  <waiting ...>

-- seg -1: con20 ==> con10, xid lock
20>: CREATE TABLE t04b (id int);  <waiting ...>

-- con20 will be cancelled by gdd
20<:  <... completed>
ERROR:  canceling statement due to user request: "cancelled by global deadlock detector"
20q: ... <quitting>

-- no more deadlock
10<:  <... completed>
UPDATE 1
10q: ... <quitting>
